### PR TITLE
[WIP] Implement client side encryption

### PR DESCRIPTION
**Pull request by NitishT**
- Added getObject() to fetch encrypted objects

- Added putObject() to put encrypted objects

### CODE REVIEW

**Comment by NitishT** [comment]
Right now only AES implementation done, still working on RSA implementation

File: api/src/main/java/io/minio/ObjectStat.java [hunk 1]
```diff
@@ -29,22 +29,30 @@
   private final long length;
   private final String etag;
   private final String contentType;
-
+  private final String key;
```
Comment (balamurugana) [inline api/src/main/java/io/minio/ObjectStat.java]: What is `key`? is it different than `name`?

File: api/src/main/java/io/minio/ObjectStat.java [hunk 2]
```diff
@@ -29,22 +29,30 @@
...
+  private final String iv;
+  private final String matDesc;
```
Comment (balamurugana) [inline api/src/main/java/io/minio/ObjectStat.java]: What are `iv` and `matDesc`?  You could give good understandable names.

File: api/src/main/java/io/minio/ObjectStat.java [hunk 3]
```diff
@@ -29,22 +29,30 @@
...
 
   /**
-   * Creates ObjectStat with given bucket name, object name, created time, object length, Etag and content type.
+   * Creates ObjectStat with given bucket name, object name, and available response header information.
    */
-  public ObjectStat(String bucketName, String name, Date createdTime, long length, String etag, String contentType) {
+  public ObjectStat(String bucketName, String name, ResponseHeader header) {
```
Comment (balamurugana) [inline api/src/main/java/io/minio/ObjectStat.java]: This change breaks backward compatibility.  You could create a new constructor.

File: api/src/main/java/io/minio/encryption/EncryptionMaterials.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	public EncryptionMaterials(SecretKey symmetricMasterKey) throws 
+	InvalidKeyException, NoSuchAlgorithmException, 
+	NoSuchPaddingException, InvalidAlgorithmParameterException{
+		
+		// Setup symmetric keys and related ciphers
+		SymmetricKey.setupSymmetricKeys(symmetricMasterKey);
+		
+	}
+	
+	public CipherInputStream decryptInputStream(InputStream encryptedInputStream, ObjectStat metadata) throws
```
Comment (balamurugana) [inline api/src/main/java/io/minio/encryption/EncryptionMaterials.java]: I could think of eliminating Object/HTTP concepts here.  Is it possible to make it as generic package where it could accept whatever info required to encrypt/decrypt Input/Output streams?

File: api/src/main/java/io/minio/ObjectStat.java (hunk 2 above)
Comment (harshavardhana) [inline api/src/main/java/io/minio/ObjectStat.java]: - x-amz-iv	Randomly generated IV (per S3 object), base64 encoded
- x-amz-matdesc	Customer provided material description in JSON (UTF8) format. Used to identify the client-side master key (ie used to encrypt/wrap the generated content encrypting key).

File: api/src/main/java/io/minio/MinioClient.java [hunk 5]
```diff
@@ -199,14 +227,21 @@ public MinioClient(String endpoint) throws InvalidEndpointException, InvalidPort
...
-   * <pre>{@code MinioClient minioClient = new MinioClient(new URL("https://play.minio.io:9000")); }</pre>
+   * </p>
+   * <b>Example:</b><br>
+   * 
+   * <pre>
+   * {
+   *   &#64;code
+   *   MinioClient minioClient = new MinioClient(new URL("https://play.minio.io:9000"));
+   * }
+   * </pre>
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: Can you send separate PR for these cleanups?

**Comment by NitishT** [comment]
Closing this. Will send new PR with changes related to only encryption changes.

**Comment by NitishT** [comment]
New PR = https://github.com/minio/minio-java/pull/561
